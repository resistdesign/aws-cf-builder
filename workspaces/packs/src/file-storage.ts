import { addParameter, createResourcePack } from '@aws-cf-builder/utils';
import { AWS, CloudFormationTemplate } from '@aws-cf-builder/types';

export type AddSecureFileStorageConfig = {
  id: string;
  bucketName?: AWS.S3.Bucket['Properties']['BucketName'] | boolean;
  label?: string;
  group?: string;
  delete?: boolean;
  blockPublicAccess?: boolean;
  cors?: AWS.S3.Bucket['Properties']['CorsConfiguration'] | boolean;
};

/**
 * Add a secure S3 Bucket with an optional parameter to set the bucket name.
 * */
export const addSecureFileStorage = createResourcePack(
  ({ id, bucketName = false, label, group, delete: shouldDelete = true, blockPublicAccess = true, cors = false }: AddSecureFileStorageConfig) => {
    const cleanLabel = !!label ? label : id;
    const ParameterId = `${id}Name`;
    const stackPatch = {
      Resources: {
        [id]: {
          Type: 'AWS::S3::Bucket',
          DeletionPolicy: shouldDelete ? 'Delete' : 'Retain',
          Properties: {
            BucketName:
              bucketName === true
                ? {
                    Ref: ParameterId,
                  }
                : bucketName === false
                ? undefined
                : bucketName,
            CorsConfiguration:
              typeof cors === 'object'
                ? cors
                : cors === true
                ? {
                    CorsRules: [
                      {
                        AllowedHeaders: ['*'],
                        AllowedMethods: ['GET', 'PUT', 'POST', 'DELETE', 'HEAD'],
                        AllowedOrigins: ['*'],
                      },
                    ],
                  }
                : undefined,
            BucketEncryption: {
              ServerSideEncryptionConfiguration: [
                {
                  ServerSideEncryptionByDefault: {
                    SSEAlgorithm: 'AES256',
                  },
                },
              ],
            },
            PublicAccessBlockConfiguration: blockPublicAccess
              ? {
                  BlockPublicAcls: true,
                  BlockPublicPolicy: true,
                  IgnorePublicAcls: true,
                  RestrictPublicBuckets: true,
                }
              : undefined,
          },
        },
      } as CloudFormationTemplate['Resources'],
    };

    return bucketName === true
      ? addParameter(
          {
            ParameterId,
            Group: !!group ? group : `${cleanLabel} Parameters`,
            Label: `${cleanLabel} Name`,
            Parameter: {
              Type: 'String',
              Description: 'The name of the file storage resource.',
            },
          },
          stackPatch as any
        )
      : stackPatch;
  }
);
