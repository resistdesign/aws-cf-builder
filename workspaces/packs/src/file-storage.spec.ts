import { addSecureFileStorage } from './file-storage';

describe('File Storage', () => {
  describe('addSecureFileStorage', () => {
    test('should add secure file storage', () => {
      const cft = addSecureFileStorage(
        {
          id: 'APIFS',
          bucketName: true,
          label: 'API FS',
          group: 'API',
          cors: true,
          blockPublicAccess: true,
          delete: true,
        },
        { AWSTemplateFormatVersion: '2010-09-09' }
      );

      expect(cft).toMatchSnapshot();
    });
  });
});
