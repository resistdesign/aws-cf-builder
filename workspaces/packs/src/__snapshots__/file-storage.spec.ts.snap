// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`File Storage addSecureFileStorage should add secure file storage 1`] = `
Object {
  "AWSTemplateFormatVersion": "2010-09-09",
  "Resources": Object {
    "APIFS": Object {
      "DeletionPolicy": "Delete",
      "Properties": Object {
        "AccessControl": undefined,
        "BucketEncryption": Object {
          "ServerSideEncryptionConfiguration": Array [
            Object {
              "ServerSideEncryptionByDefault": Object {
                "SSEAlgorithm": "AES256",
              },
            },
          ],
        },
        "BucketName": true,
        "CorsConfiguration": Object {
          "CorsRules": Array [
            Object {
              "AllowedHeaders": Array [
                "*",
              ],
              "AllowedMethods": Array [
                "GET",
                "PUT",
                "POST",
                "DELETE",
                "HEAD",
              ],
              "AllowedOrigins": Array [
                "*",
              ],
            },
          ],
        },
        "PublicAccessBlockConfiguration": Object {
          "BlockPublicAcls": true,
          "BlockPublicPolicy": true,
          "IgnorePublicAcls": true,
          "RestrictPublicBuckets": true,
        },
      },
      "Type": "AWS::S3::Bucket",
    },
  },
}
`;
